HPCC-31196 Add "disconnect", "unlock", etc. to WsDali

wangkx opened the pull request:
Add "disconnect", "unlock", "settracetransactions", "settraceslowtransactions", "cleartracetransactions", "save" (from dalidiag) to the WsDali service.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-31196 Add "disconnect", "unlock", etc. to WsDali

Add "disconnect", "unlock", "settracetransactions",
"settraceslowtransactions", "cleartracetransactions",
"save" (from dalidiag) to the WsDali service.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented on esp/scm/ws_dali.ecm:
```diff
@@ -151,8 +151,36 @@ ESPrequest [nil_remove] GetSDSSubscribersRequest
 
+ESPrequest [nil_remove] DisconnectClientConnectionRequest
+{
+    string URL;
```
a url is misleading. It's a host:port. Endpoint would be better.

jakesmith commented on esp/services/ws_dali/ws_daliservice.cpp:
```diff
@@ -594,3 +594,135 @@ bool CWSDaliEx::onGetSDSSubscribers(IEspContext& context, IEspGetSDSSubscribersR
+
+        const char* url = req.getURL();
+        if (isEmptyString(url))
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "URL not specified.");
```
see prev. comment re. "endpoint"

jakesmith commented on esp/services/ws_dali/ws_daliservice.cpp:
```diff
@@ -594,3 +594,135 @@ bool CWSDaliEx::onGetSDSSubscribers(IEspContext& context, IEspGetSDSSubscribersR
+            throw makeStringException(ECLWATCH_INVALID_INPUT, "ConnectionID not specified.");
+
+        MemoryBuffer mb;
+        mb.append("unlock").append(strtoll(connectionIdHex, nullptr, 16)).append(req.getClose());
```
I had to check this, because it's not clear from the dalidiag usage (not a new issue) that it expects a hex number.
I think worth updating the daliadmin usage text at the same time to note it expect the connectionID to be hex.

jakesmith reviewed (commented):
@wangkx - looks good, just a couple of minor comments.

wangkx committed:
HPCC-31196 Revise based on review

1. Inside the DisconnectClientConnectionRequest, rename the URL
to Endpoint.
2. Update the usage text for dalidiag -unlock.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review my changes and let me know if I miss any.

jakesmith reviewed (approved):
@wangkx - looks fine. Please squash.
